packages/server-boilerplate/src/utils/initialiseApiClient.ts

```diff
@@ -0,0 +1,170 @@
+import { encryptPassword } from '@tupaia/auth';
+import { ServerBoilerplateModelRegistry } from '../types';
+
+const upsertUserAccount = async ({
+  models,
+  email,
+  password,
+  salt,
+}: {
+  models: ServerBoilerplateModelRegistry;
+  email: string;
+  password: string;
+  salt: string;
+}): Promise<string> => {
+  const passwordHash = encryptPassword(password, salt);
+  const firstName = email;
+  const lastName = 'API Client';
+
+  const existingUserAccount = await models.user.findOne({ email: email });
+
+  if (existingUserAccount) {
+    await models.user.update(
+      { email: email },
+      {
+        first_name: firstName,
+        last_name: lastName,
+        email: email,
+        password_hash: passwordHash,
+        password_salt: salt,
+      },
+    );
+    return existingUserAccount.id;
+  }
+
+  const newId = generateId();
+  await models.user.create({
+    id: newId,
+    first_name: firstName,
+    last_name: lastName,
+    email: email,
+    password_hash: passwordHash,
+    password_salt: salt,
+  });
+  return newId;
+};
+
+const upsertApiClient = async ({
+  models,
+  userAccountId,
+  username,
+  password,
+  salt,
+}: {
+  models: ServerBoilerplateModelRegistry;
+  userAccountId: string;
+  username: string;
+  password: string;
+  salt: string;
+}) => {
+  const secretKeyHash = encryptPassword(password, salt);
+
+  const existingApiClient = await models.apiClient.findOne({
+    username: username,
+  });
+
+  if (existingApiClient) {
+    await models.apiClient.update(
+      { username: username },
+      {
+        secret_key_hash: secretKeyHash,
+        user_account_id: userAccountId,
+      },
+    );
+    return;
+  }
+  await models.apiClient.create({
+    id: generateId(),
+    username: username,
+    secret_key_hash: secretKeyHash,
+    user_account_id: userAccountId,
+  });
+};
+
+const upsertPermissions = async ({
+  models,
+  userAccountId,
+  permissions,
+}: {
+  models: ServerBoilerplateModelRegistry;
+  userAccountId: string;
+  permissions: { entityCode: string; permissionGroupName: string }[];
+}) => {
+  await models.userEntityPermission.delete({ user_id: userAccountId });
+
+  const entities = await models.entity.find({ code: permissions.map(p => p.entityCode) });
+  const entityIdByCode = Object.fromEntries(entities.map(entity => [entity.code, entity.id]));
+
+  const permissionGroups = await models.permissionGroup.find({
+    name: permissions.map(p => p.permissionGroupName),
+  });
+  const permissionGroupIdByName = Object.fromEntries(
+    permissionGroups.map(permissionGroup => [permissionGroup.name, permissionGroup.id]),
+  );
+
+  await models.userEntityPermission.createMany(
+    permissions.map(p => ({
+      id: generateId(),
+      user_id: userAccountId,
+      entity_id: entityIdByCode[p.entityCode],
+      permission_group_id: permissionGroupIdByName[p.permissionGroupName],
+    })),
+  );
+
+  await models.userEntityPermission.createMany(
```

Comment:
I might be being a bit dense here, but is this the same as above?

Reply:
Not dense!! Good catch!